### PR TITLE
Implement IDL based generators

### PR DESCRIPTION
**Pull request by Diggsey**
Example generated bindings:
https://gist.github.com/Diggsey/c8a4cb93c7d5643d20203401ead178a7

This can't be merged until:
1) I've tested it beyond the bindings compiling
2) The next version of stdweb is released, which will contain the `CanvasElement` type

### CODE REVIEW

**Comment by Diggsey** [comment]
I've just implemented a massive rework of the type-generation - it's now really powerful, and will generate borrowed versions of types for arguments and will generate generic methods when appropriate.

**Comment by Diggsey** [comment]
image

**Comment by brendanzab** [comment]
Woo! This is looking great!

**Comment by brendanzab** [comment]
I'm wondering if it might actually make sense to create a new crate instead under this repository, called `webgl_generator`?

**Comment by Diggsey** [comment]
Yeah, I don't really mind, is that what you'd prefer?

**Comment by brendanzab** [comment]
Wasn't talking about creating a new repo, just a new crate as a directory in this repo. If I'm not mistaken it seems like `gl_generator` and `webgl_generator` are orthogonal in their implementations and their dependencies?

**Comment by tomaka** [comment]
Oh sorry, yes that would make sense!

**Comment by edwin0cheng** [comment]
@Diggsey As i mentioned in https://github.com/koute/stdweb/issues/92, will it works in Firefox ?

**Comment by Diggsey** [comment]
@edwin0cheng not right now - at least not if you try do anything complex. I think this can (and should) be fixed from the stdweb side though.

**Comment by brendanzab** [comment]
Looking great so far! Let me know when this can progress forward!

**Comment by Diggsey** [comment]
@brendanzab I've realised that the IDL files are missing information on extensions - these are described via separate IDL embedded in XML in the Khronos WebGL repo.

I would like to expose a mapping from "<extension name>" => "<IDL contents>" in the `khronos_api` crate - this will involve a build step to extract the IDL from the XML and will need to be stores as a static array of type `&'static [(&'static str, &'static str)]`. Does that sound reasonable to you?

**Comment by Diggsey** [comment]
I've opened a separate PR to add extension support to `khronos_api`, and I've rebased this PR on top.

The generated bindings are now up to 10,000 lines, and include support for all accepted WebGL extensions. There's also a "version-less" `GLContext` type to make it easier to write code which can work with multiple versions.

I've uploaded the generated documentation as I think it's quite interesting to read:
http://awesome-davinci-d3197a.bitballoon.com/webgl_stdweb/index.html

**Comment by brendanzab** [comment]
> I would like to expose a mapping from "<extension name>" => "<IDL contents>" in the khronos_api crate - this will involve a build step to extract the IDL from the XML and will need to be stored as a static array of type &'static [(&'static str, &'static str)]. Does that sound reasonable to you?

Oh, missed this comment! Would have been helpful if you had have linked to it in the PR of #451! Yes, that sounds great.

**Comment by Diggsey** [comment]
One thing that might be cool would be to pull out the "summary" and "function" sections from the XML files and attach them as doc-comments to the generated code.

**Comment by brendanzab** [comment]
Ooooh! Wondering if we could do that for all the generators 😮

**Comment by Diggsey** [comment]
@brendanzab stdweb 0.4.0 was released, and I've update the PR to use that. It should be ready for you to review now!

**Comment by brendanzab** [comment]
Woo, looking now!

File: webgl_generator/webgl_generators/stdweb_gen.rs [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+        ProcessedResult { type_: name.into(), wrapper: ResultWrapper::TryInto, optional: false }
+    }
+}
+
+fn process_result_type_kind(name: &str, type_kind: &TypeKind, registry: &Registry) -> ProcessedResult {
+    match type_kind {
+        &TypeKind::Primitive(ref p) => ProcessedResult::simple(p.name()),
+        &TypeKind::String => ProcessedResult::simple("String"),
+        &TypeKind::ArrayBuffer | &TypeKind::ArrayBufferView => ProcessedResult::simple("ArrayBuffer"),
+        &TypeKind::BufferSource => unimplemented!("BufferSource not supported in output"),
```
Comment (brendanzab) [inline webgl_generator/webgl_generators/stdweb_gen.rs]: Should we be using `unimplemented!` here? Are you planning to add support for this later?

File: webgl_generator/webgl_generators/stdweb_gen.rs [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+
+    let args: Vec<_> = operation.args.iter().map(|a| {
+        let processed = process_arg_type(&a.type_, registry, &mut gc);
+        let arg = format!("{}: {}", unreserve(snake(&a.name)), processed.type_);
+        let js_arg = format!("@{{{}}}", processed.wrapper.wrap(&unreserve(snake(&a.name))));
+        OperationArg { arg, js_arg }
+    }).collect();
+
+    let rust_args = args.iter().map(|a| a.arg.clone()).collect::<Vec<_>>().join(", ");
+    let js_args = args.iter().map(|a| a.js_arg.clone()).collect::<Vec<_>>().join(", ");
```
Comment (brendanzab) [inline webgl_generator/webgl_generators/stdweb_gen.rs]: itertools has [`join`](https://docs.rs/itertools/0.7.6/itertools/fn.join.html) if you wanted to use that to skip the intermediate `collect`!

File: webgl_generator/webgl_registry/mod.rs [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+            SignedLong => TypeKind::Primitive(Primitive::I32),
+            UnsignedLong => TypeKind::Primitive(Primitive::U32),
+            SignedLongLong => TypeKind::Primitive(Primitive::I64),
+            UnsignedLongLong => TypeKind::Primitive(Primitive::U64),
+            RestrictedFloat | UnrestrictedFloat => TypeKind::Primitive(Primitive::F32),
+            RestrictedDouble | UnrestrictedDouble => TypeKind::Primitive(Primitive::F64),
+
+            // Strings
+            DOMString | USVString => TypeKind::String,
+            ByteString => unimplemented!(),
```
Comment (brendanzab) [inline webgl_generator/webgl_registry/mod.rs]: Has this been forgotten?

File: webgl_generator/webgl_registry/mod.rs [hunk 4]
```diff
@@ -0,0 +1,779 @@
...
+        let type_ = self.load_type(*typedef.type_);
+        self.load_type_kind(&typedef.name, TypeKind::Typedef(type_));
+    }
+
+    fn load_field(&mut self, field: ast::DictionaryMember) -> Option<(String, Field)> {
+        let type_ = self.load_type(*field.type_);
+
+        Some((field.name, Field {
+            type_
+        }))
```
Comment (brendanzab) [inline webgl_generator/webgl_registry/mod.rs]: Why does this need to be an `Option`?

**Comment by brendanzab** [comment]
Would be nice to Run rustfmt on this if you don't mind!

File: webgl_generator/webgl_generators/stdweb_gen.rs (hunk 1 above)
Comment (Diggsey) [reply webgl_generator/webgl_generators/stdweb_gen.rs]: `BufferSource` is not currently used in the WebGL IDL files, so I haven't bothered implementing it.

**Comment by Diggsey** [comment]
@brendanzab done

**Comment by Diggsey** [comment]
@brendanzab hey is this good to go?

**Comment by brendanzab** [comment]
Ack - left this one sit around in my notifications making me feel guilty for too long! Merging!

**Comment by Diggsey** [comment]
Thanks! Are you happy to publish the new crate(s) to crates.io?

**Comment by brendanzab** [comment]
Yup, will do. I'll add you to the crate owners too.

**Comment by brendanzab** [comment]
Hum, when publishing I get:

```
error[E0425]: cannot find value `WEBGL_EXT_XML` in module `khronos_api`
  --> webgl_registry/mod.rs:69:38
   |
69 |         for &ext_xml in khronos_api::WEBGL_EXT_XML {
   |                                      ^^^^^^^^^^^^^ not found in `khronos_api`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
error: failed to verify package tarball

Caused by:
  Could not compile `webgl_generator`.
```

**Comment by brendanzab** [comment]
Oh, I think I need to bump the khronos_api version 🤔

**Comment by brendanzab** [comment]
Oh, and probably also publish those changes

**Comment by Diggsey** [comment]
Yes, I think so?

**Comment by brendanzab** [comment]
Done!

**Pull request merged**